GVFS/GVFS.Common/Tracing/JsonTracer.cs

```diff
@@ -267,10 +291,45 @@ private void WriteEvent(string eventName, EventLevel level, Keywords keywords, E
                 throw new ObjectDisposedException(nameof(JsonTracer));
             }
 
+            var message = new TraceEventMessage
+            {
+                EventName = eventName,
+                ActivityId = this.activityId,
+                ParentActivityId = this.parentActivityId,
+                Level = level,
+                Keywords = keywords,
+                Opcode = opcode,
+                Payload = jsonPayload
+            };
+
             foreach (EventListener listener in this.listeners)
             {
-                listener.RecordMessage(eventName, this.activityId, this.parentActivityId, level, keywords, opcode, jsonPayload);
+                string errorMessage;
+                if (!listener.TryRecordMessage(message, out errorMessage))
+                {
+                    this.MarkAndLogListenerFailure(listener, errorMessage);
+                }
+            }
+        }
+
+        private void MarkAndLogListenerFailure(EventListener failingListener, string errorMessage)
+        {
+            if (this.failedListeners.ContainsKey(failingListener))
+            {
+                // We've already logged that this listener has failed so there is no need to do it again
+                return;
+            }
+
+            this.failedListeners.Add(failingListener, errorMessage);

```

Comment:
When does the failed listener get removed from the failed list if the next message is successful?  For example the service daemon is running again.

Reply:
Currently it does not, but logging the recovery (only once after a successful message on a previously failing listener) might be useful – I'll add it.